components/style/selector_impl.rs

```diff
@@ -8,6 +8,73 @@ use selectors::parser::{ParserContext, SelectorImpl};
 pub enum PseudoElement {
     Before,
     After,
+    FirstLine,
+
+    // https://mxr.mozilla.org/mozilla-central/source/layout/style/nsCSSAnonBoxList.h
```

Comment:
> Would it be fine to have a feature-gated gecko selector implementation inside style?

I’m not opposed but I don’t see how it would help.

> Becomes this: [...]

Yes, calling parsing code here is silly. It shows a mis-match between `PseudoElement` being generic and the `before_map` and `after_map` fields of `Stylist` being hard-coded. I don’t think this means me should abandon the generic parser and have a single `PseudoElement` type. I’d rather change the way `Stylist` is organized. Maybe some map of `PseudoElement` to `PerPseudoElementSelectorMap`?